azure-iot-device/tests/iothub/transport/mqtt/test_mqtt_provider.py

```diff
@@ -8,164 +8,1120 @@
+
+    @pytest.mark.it("Starts MQTT Network Loop")
+    def test_calls_loop_start(self, mocker, mock_mqtt_client, provider):
+        provider.connect(fake_password)
+
+        assert mock_mqtt_client.loop_start.call_count == 1
+        assert mock_mqtt_client.loop_start.call_args == mocker.call()
+
+    @pytest.mark.it("Triggers on_mqtt_connected event handler callback upon connect completion")
+    def test_calls_event_handler_callback(self, mocker, mock_mqtt_client, provider):
+        callback = mocker.MagicMock()
+        provider.on_mqtt_connected = callback
+
+        # Initiate connect
+        provider.connect(fake_password)
+
+        # Manually trigger Paho on_connect event_handler
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # Verify provider.on_mqtt_connected was called
+        assert callback.call_count == 1
+        assert callback.call_args == mocker.call()
+
+    @pytest.mark.it(
+        "Skips on_mqtt_connected event handler callback if set to 'None' upon connect completion"
+    )
+    def test_skips_none_event_handler_callback(self, mocker, mock_mqtt_client, provider):
+        assert provider.on_mqtt_connected is None
+
+        provider.connect(fake_password)
+
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # No further asserts required - this is a test to show that it skips a callback.
+        # Not raising an exception == test passed
+
+    @pytest.mark.it("Recovers from exception in on_mqtt_connected event handler callback")
+    def test_event_handler_callback_raises_exception(self, mocker, mock_mqtt_client, provider):
+        event_cb = mocker.MagicMock(side_effect=DummyException)
+        provider.on_mqtt_connected = event_cb
+
+        provider.connect(fake_password)
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # Callback was called, but exception did not propagate
+        assert event_cb.call_count == 1
+
+
+@pytest.mark.describe("MQTT Provider - Reconnect")
+class TestReconnect(object):
+    @pytest.mark.it("Sets username and password")
+    def test_sets_username_and_password(self, mocker, mock_mqtt_client, provider):
+        provider.reconnect(fake_password)
+
+        assert mock_mqtt_client.username_pw_set.call_count == 1
+        assert mock_mqtt_client.username_pw_set.call_args == mocker.call(
+            username=fake_username, password=fake_password
+        )
+
+    @pytest.mark.it("Reconnects with Paho")
+    def test_calls_paho_reconnect(self, mocker, mock_mqtt_client, provider):
+        provider.reconnect(fake_password)
+
+        assert mock_mqtt_client.reconnect.call_count == 1
+        assert mock_mqtt_client.reconnect.call_args == mocker.call()
+
+    @pytest.mark.it(
+        "Triggers on_mqtt_connected event handler callback upon user-driven reconnect completion"
```

Comment:
>upon user-driven reconnect completion [](start = 59, length = 37)

I'm not sure if the phrase "user-driven" belongs in here.  The test is simulating a reconnect completion, but I wouldn't call that "user-driven".

Reply:
ah-ha,  it should say "upon completion of user-driven reconnect" to show that the reconnect is user-driven.  With the existing wording, I though the completion was user-driven

---
In reply to: [277847664](https://github.com/Azure/azure-iot-sdk-python-preview/pull/82#discussion_r277847664) [](ancestors = 277847664)

Reply:
Good point. I'll make this change here and everywhere.

---
In reply to: [277848527](https://github.com/Azure/azure-iot-sdk-python-preview/pull/82#discussion_r277848527) [](ancestors = 277848527,277847664)